R/read.R

```diff
@@ -289,23 +311,87 @@ readH5AD <- function(file, X_name = NULL, use_hdf5 = FALSE,
+        codes <- obj[["codes"]] + 1
+        codes[codes == 0] <- NA
+        levels <- obj[["categories"]]
+
+        # HACK rhdf5 doesn't yet support enums in attrs
+        # (h5readAttributes() above may warn about this)
+        #ord <- as.logical(element_attrs[["ordered"]])
+        ord <- FALSE
+
+        obj <- factor(levels[codes], levels=levels, ordered=ord)
+        return(obj)
+    }
+
+    # Handle nullable booleans/integers
+    if (element_attrs[["encoding-type"]] %in% c("nullable-boolean",
+                                                "nullable-integer")) {
+        mask <- as.logical(obj[["mask"]]) # convert enum to bool
+        obj <- obj[["values"]]
+        obj[mask] <- NA
+    }
+
+    # Handle booleans. Non-nullable booleans have encoding-type
+    # "array", so we have to infer the type from the enum levels
+    if (is.factor(obj) && identical(levels(obj), c("FALSE", "TRUE"))) {
+        obj <- as.logical(obj)
+        return(obj)
+    }
+
+    # Recursively convert element members
+    if (recursive && is.list(obj) && !is.null(names(obj))) {
+        for (k in names(obj)) {
+            obj[[k]] <- .convert_element(
+                obj[[k]], file.path(path, k),
+                file, recursive=TRUE
+            )
+        }
+    }
+
+    if (is.list(obj) && !is.null(names(obj))) {
+        names(obj) <- make.names(names(obj))
+    }
+
+    obj
+}
+
+#' @importFrom utils compareVersion
 #' @importFrom S4Vectors DataFrame
-.read_dim_data <- function(file, path, fields) {
+.read_dim_data <- function(file, path, fields, version) {
     col_names <- setdiff(names(fields), c("__categories", "_index"))
     out_cols <- list()
     for (col_name in col_names) {
-        out_cols[[col_name]] <- as.vector(
-            rhdf5::h5read(file, file.path(path, col_name))
-        )
+        vec <- rhdf5::h5read(file, file.path(path, col_name))
+
+        if (compareVersion(version, "0.8") >= 0) {
+            vec <- .convert_element(
+                vec, file.path(path, col_name),
+                file, recursive=FALSE
+            )
+        }
+
+        if (!is.factor(vec)) {
+            vec <- as.vector(vec)
+        }
+
+        out_cols[[col_name]] <- vec
     }
 
-    cat_names <- names(fields[["__categories"]])
-    for (cat_name in cat_names) {
-        levels <- as.vector(
-            rhdf5::h5read(file, file.path(path, "__categories", cat_name))
-        )
-        out_cols[[cat_name]] <- factor(out_cols[[cat_name]])
-        levels(out_cols[[cat_name]]) <- levels
+    if (compareVersion(version, "0.8") < 0) {
+        cat_names <- names(fields[["__categories"]])
+        for (cat_name in cat_names) {
+            levels <- as.vector(
+                rhdf5::h5read(file, file.path(path, "__categories", cat_name))
+            )
+            out_cols[[cat_name]] <- factor(out_cols[[cat_name]])
+            levels(out_cols[[cat_name]]) <- levels
+        }
```

Comment:
This is the problem I mentioned in my comment just now. AnnData 0.7 categories are only converted if `version` explicitly specified.